app/uk/gov/hmrc/hec/controllers/IFController.scala

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.hec.controllers
+
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+import java.util.UUID
+
+import com.google.inject.{Inject, Singleton}
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import uk.gov.hmrc.hec.models.TaxYear
+import uk.gov.hmrc.hec.models.ids.{CTUTR, SAUTR}
+import uk.gov.hmrc.hec.services.{IFService, IFServiceImpl}
+import uk.gov.hmrc.hec.util.Logging
+import uk.gov.hmrc.hec.util.Logging.LoggerOps
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+import cats.data.{NonEmptyList, Validated, ValidatedNel}
+import cats.data.Validated._
+import cats.implicits._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class IFController @Inject() (
+  IFService: IFService,
+  cc: ControllerComponents
+)(implicit ec: ExecutionContext)
+    extends BackendController(cc)
+    with Logging {
+
+  private def correlationId = UUID.randomUUID().toString
+
+  /**
+    * Fetch individual user's self-assessment status
+    * @param utr Self-assessment UTR
+    * @param taxYear Start tax year
+    * @return Self-assessment status with UTR and tax year
+    */
+  def getSAStatus(utr: String, taxYear: String): Action[AnyContent] = Action.async { implicit request =>
+    val sautrValidation   = SAUTR.fromString(utr).toValidNel("Invalid SAUTR")
+    val taxYearValidation = TaxYear.fromString(taxYear).toValidNel("Invalid tax year")
+
+    @SuppressWarnings(Array("org.wartremover.warts.Any"))
+    val validation = (sautrValidation, taxYearValidation).mapN((_, _))
+    validation match {
+      case Valid((utr, year)) =>
+        IFService
+          .getSAStatus(utr, year, correlationId)
+          .fold(
+            {
+              case IFServiceImpl.DataError(msg)  =>
+                logger.warn(s"[correlationId:$correlationId] $msg")
```

Comment:
won't this log a random UUID that's different than the one passed into the service? It's using `correlationId` which is a def. What probably needs to happen is a correlation id needs to be generated at the start of this function and stored in a `val` and used where it's needed. `correlationId` maybe then becomes `newCorrelationId` or something so that you can name the val `correlationId`

Reply:
🤦 you are right, I'll fix

Reply:
updated in commit - https://github.com/hmrc/hec/pull/3/commits/202a90e00fd544187f315c7ee45bf342911da272

Reply:
I've used this format to include the correlationId in the logs `[correlationId:12345]...`